### PR TITLE
Update generated client design

### PR DESCRIPTION
**Pull request by kaneeldias**
## Purpose
$subject

New design example: https://github.com/sameerajayasoma/bal-persist-design-prototypes/tree/main/rainier-sample

Fixes: https://github.com/ballerina-platform/ballerina-standard-library/issues/3927

## Checklist
- [x] Linked to an issue
- [ ] Updated the specification
- [ ] Updated the changelog
- [x] Added tests

### CODE REVIEW

File: ballerina/tests/persist/rainier.bal [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    string lastName;
+    time:Date birthDate;
+    string gender;
+    time:Date hireDate;
+
+    Department department;
+    Workspace workspace;
+|};
+
+table<Employee> key(empNo) employees = table[];
```
Comment (niveathika) [inline ballerina/tests/persist/rainier.bal:54]: Removed from latest design

File: ballerina/tests/persist/rainier.bal [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import ballerina/time;
```
Comment (niveathika) [inline ballerina/tests/persist/rainier.bal:1]: IMO these tests should come as separate modules. Lets do the refactoring in a separate PR

**Review by niveathika — APPROVED** [review]
LGTM. Lets address comments in separate PR

File: ballerina/tests/persist/rainier.bal (hunk 1 above)
Comment (kaneeldias) [reply ballerina/tests/persist/rainier.bal]: Fixed in [5d9b949](https://github.com/ballerina-platform/module-ballerina-persist/pull/103/commits/5d9b94991b34f96099ce43fb3e0987b7f02dcaa9)

File: ballerina/sql_client.bal [hunk 3]
```diff
@@ -249,20 +280,15 @@ public client class SQLClient {
...
 
-    # Closes the underlying `sql:Client`.
-    #
-    # + return - `()` if the client is closed successfully or a `persist:Error` if the operation fails
-    public isolated function close() returns Error? {
-        sql:Error? e = self.dbClient.close();
-        if e is sql:Error {
-            return <Error>error(e.message());
-        }
-    }
```
Comment (daneshk) [inline ballerina/sql_client.bal:260]: why we remove the close function

File: ballerina/tests/generated_client.bal [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+                    record {|Building value;|} nextRecord = {value: check streamValue.value.cloneWithType(Building)};
+                    return nextRecord;
+                } on fail error e {
+                    return <Error>e;
+                }
+            }
+        } else {
+            // Unreachable code
+            return ();
+        }
```
Comment (niveathika) [inline ballerina/tests/generated_client.bal:230]: Have you looked into the possibility to pass the typedescription to a common function? We can reduce a whole lot of code if we do the above

File: ballerina/tests/generated_client.bal [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+    }
+
+    public isolated function close() returns Error? {
+        if self.anydataStream is stream<anydata, sql:Error?> {
```
Comment (niveathika) [inline ballerina/tests/generated_client.bal:234]: Same

File: ballerina/tests/generated_types.bal [hunk 6]
```diff
@@ -0,0 +1,74 @@
+import ballerina/time;
+
+// TODO: Remove "public" from the generated types
+public enum Gender {
+    M,
+    F
+}
```
Comment (niveathika) [inline ballerina/tests/generated_types.bal:7]: Remove

File: ballerina/tests/generated_types.bal [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    string city;
+    string state;
+    string country;
+    string postalCode;
+|};
+
+public type Workspace record {|
+    readonly string workspaceId;
+    string workspaceType;
+    string buildingCode;
```
Comment (niveathika) [inline ballerina/tests/generated_types.bal:21]: Is this the correct generated models? Shouldint the foreign key be `buildingBuildingCode`?

File: examples/medical-center/persist/medical_center.bal [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import ballerina/time;
```
Comment (niveathika) [inline examples/medical-center/persist/medical_center.bal:1]:
```suggestion
import ballerina/time;
import ballerina/persist as _;
```

File: examples/medical-center/generated_client.bal [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    };
+
+    private final map<persist:SQLClient> persistClients;
+
+    public function init() returns persist:Error? {
+        do {
+            self.dbClient = check new (host = host, user = user, password = password, database = database, port = port);
+
+            self.persistClients = {
+                medicalItem: check new (self.dbClient, self.metadata.get("medicalItem").entityName, self.metadata.get("medicalItem").tableName, self.metadata.get("medicalItem").keyFields, self.metadata.get("medicalItem").fieldMetadata),
```
Comment (sahanHe) [inline examples/medical-center/generated_client.bal:44]: Shouldn't these be string types?

File: examples/medical-center/generated_client.bal [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+import ballerina/persist;
+
+client class MedicalCenterClient {
+
+    private final mysql:Client dbClient;
+
+    private final map<persist:Metadata> metadata = {
+        medicalItem: {
+            entityName: "MedicalItem",
+            tableName: `MedicalItem`,
```
Comment (sahanHe) [inline examples/medical-center/generated_client.bal:13]: I this format for table name correct?

Comment (sahanHe) [reply examples/medical-center/generated_client.bal:13]: Resolved

File: examples/medical-center/persist/medical_center.bal (hunk 8 above)
Comment (kaneeldias) [reply examples/medical-center/persist/medical_center.bal:1]: Fixed in [90235d8](https://github.com/ballerina-platform/module-ballerina-persist/pull/103/commits/90235d817df86c78d05ff4528baba5a18d2a584a)

File: ballerina/tests/generated_types.bal (hunk 7 above)
Comment (kaneeldias) [reply ballerina/tests/generated_types.bal]: Fixed in [90235d8](https://github.com/ballerina-platform/module-ballerina-persist/pull/103/commits/90235d817df86c78d05ff4528baba5a18d2a584a)

File: ballerina/tests/generated_types.bal (hunk 6 above)
Comment (kaneeldias) [reply ballerina/tests/generated_types.bal]: Fixed in [90235d8](https://github.com/ballerina-platform/module-ballerina-persist/pull/103/commits/90235d817df86c78d05ff4528baba5a18d2a584a)

File: ballerina/tests/generated_client.bal (hunk 5 above)
Comment (kaneeldias) [reply ballerina/tests/generated_client.bal]: Fixed in [90235d8](https://github.com/ballerina-platform/module-ballerina-persist/pull/103/commits/90235d817df86c78d05ff4528baba5a18d2a584a)

File: ballerina/tests/generated_client.bal [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+                department: check new (self.dbClient, self.metadata.get("department").entityName, self.metadata.get("department").tableName, self.metadata.get("department").keyFields, self.metadata.get("department").fieldMetadata),
+                employee: check new (self.dbClient, self.metadata.get("employee").entityName, self.metadata.get("employee").tableName, self.metadata.get("employee").keyFields, self.metadata.get("employee").fieldMetadata)
+            };
+        } on fail error e {
+            return <Error>error(e.message());
+        }
+    }
+
+    isolated resource function get buildings() returns stream<Building, error?> {
+        stream<anydata, sql:Error?>|Error result = self.persistClients.get("building").runReadQuery(Building);
```
[the remark this answers is not in the document]
Comment (kaneeldias) [reply ballerina/tests/generated_client.bal]: Fixed in [90235d8](https://github.com/ballerina-platform/module-ballerina-persist/pull/103/commits/90235d817df86c78d05ff4528baba5a18d2a584a)

File: ballerina/tests/generated_client.bal [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+    };
+    
+    private final map<SQLClient> persistClients;
+    
+    public function init() returns Error? {
+        do {
+            self.dbClient = check new (host = host, user = user, password = password, database = database, port = port);
+
+            self.persistClients = {
+                building: check new (self.dbClient, self.metadata.get("building").entityName, self.metadata.get("building").tableName, self.metadata.get("building").keyFields, self.metadata.get("building").fieldMetadata),
```
[the remark this answers is not in the document]
Comment (kaneeldias) [reply ballerina/tests/generated_client.bal]: Fixed in [90235d8](https://github.com/ballerina-platform/module-ballerina-persist/pull/103/commits/90235d817df86c78d05ff4528baba5a18d2a584a)

File: ballerina/tests/generated_client.bal [hunk 13]
```diff
@@ -0,0 +1,374 @@
+import ballerinax/mysql;
+import ballerina/sql;
+import ballerina/time;
+
+client class RainierClient {
+
+    private final mysql:Client dbClient;
+
+    private final map<Metadata> metadata = {
```
[the remark this answers is not in the document]
Comment (kaneeldias) [reply ballerina/tests/generated_client.bal]: Fixed in [90235d8](https://github.com/ballerina-platform/module-ballerina-persist/pull/103/commits/90235d817df86c78d05ff4528baba5a18d2a584a)

File: ballerina/tests/generated_client.bal (hunk 4 above)
Comment (kaneeldias) [reply ballerina/tests/generated_client.bal:235]: As discussed offline, this is currently not alllowed

**Review by daneshk — APPROVED** [review]

**Comment by daneshk** [comment]
@niveathika please approve and merge

**Review by niveathika — APPROVED** [review]
LGTM
